riscv/csrs.cc

```diff
@@ -620,20 +620,49 @@ reg_t misa_csr_t::dependency(const reg_t val, const char feature, const char dep
 }
 
 bool misa_csr_t::unlogged_write(const reg_t val) noexcept {
-  // the write is ignored if increasing IALIGN would misalign the PC
-  if (!(val & (1L << ('C' - 'A'))) && (state->pc & 2))
-    return false;
-
   reg_t adjusted_val = val;
   adjusted_val = dependency(adjusted_val, 'D', 'F');
   adjusted_val = dependency(adjusted_val, 'Q', 'D');
   adjusted_val = dependency(adjusted_val, 'V', 'D');
 
   const reg_t old_misa = read();
-  const bool prev_h = old_misa & (1L << ('H' - 'A'));
   const reg_t new_misa = (adjusted_val & write_mask) | (old_misa & ~write_mask);
-  const bool new_h = new_misa & (1L << ('H' - 'A'));
 
+  const bool prev_c = old_misa & (1L << ('C' - 'A'));
+  const bool new_c = new_misa & (1L << ('C' - 'A'));
+  if (!new_c && prev_c) {
+    if (!(proc->extension_enabled(EXT_ZCB) ||
+          proc->extension_enabled(EXT_ZCMP) ||
+          proc->extension_enabled(EXT_ZCMT))) {
+      // the write is ignored if increasing IALIGN would misalign the PC
+      if (state->pc & 2)
+        return false;
+      proc->disable_extension(EXT_ZCA);
+      proc->disable_extension(EXT_ZCF);
+      proc->disable_extension(EXT_ZCD);
+    }
+  } else if (new_c && !prev_c) {
+    proc->enable_extension(EXT_ZCA);
+  }
+
+  const bool prev_f = old_misa & (1L << ('F' - 'A'));
+  const bool new_f = new_misa & (1L << ('F' - 'A'));
+  if (!new_f && prev_f) {
+    proc->disable_extension(EXT_ZCF);
+  } else if (new_f && !prev_f && proc->extension_enabled(EXT_ZCA)) {
+    proc->enable_extension(EXT_ZCF);
+  }
```

Comment:
I assume Zca requires Zcf? This is not the best place to express that restriction, and I suspect it's a duplicate of the ISA parser which also checks that restriction.

If the ISA object will not allow Zcf without Zca then I don't see the harm in calling `enable_extension` here, since `isa->extension_enabled(EXT_ZCF)` will still be false.

This entire `if` should be one function call: `set_extension_enable(EXT_ZCF, new_f)`

Reply:
Zcf requires Zca. We cannot enable Zcf without Zca enabled. The corner case is  isa=CF, if we disable C and F firstly, and enable F without C later.

Reply:
This is a really confusing block of code. We need a way to express these dependencies succinctly instead of through jumbled logic like this. For example we have `misa_csr_t::dependency()` which enables understandable adjustments of the `misa` value like so:
```
  adjusted_val = dependency(adjusted_val, 'D', 'F');
  adjusted_val = dependency(adjusted_val, 'Q', 'D');
  adjusted_val = dependency(adjusted_val, 'V', 'D');
```